src/Polysemy/Internal/Writer.hs

```diff
@@ -205,17 +205,11 @@ interpretViaLazyWriter f sem = Sem $ \(k :: forall x. Union r (Sem r) x -> m x)
   let
     go :: forall x. Sem (e ': r) x -> Lazy.WriterT o m x
     go = usingSem $ \u -> case decomp u of
-      Right (Weaving e s wv ex ins) -> f $ Weaving e s (go . wv) ex ins
-      Left g -> Lazy.WriterT $ do
-        ~(o, a) <- k $
-          weave
-            (mempty, ())
-            (\ ~(o, m) -> (fmap . first) (o <>) (interpretViaLazyWriter f m))
-            (Just . snd)
-            g
-        return (a, o)
+      Right (Weaving e mkT lwr ex) -> f $ Weaving e (\n -> mkT (n . go)) lwr ex
+      Left g ->
+        liftHandlerWithNat
+          (Lazy.WriterT . fmap swap . interpretViaLazyWriter f)
+          k g
     {-# INLINE go #-}
-  in do
-    ~(a,s) <- Lazy.runWriterT (go sem)
```

Comment:
I think this was purposefully lazier than `swap` initially, for some performance reasons. I don't know if they are relevant (I think Sandy even had a specialised `swap` for this purpose at some point?)

Reply:
You're right on these being purposefully lazier swaps, but not as to why. [Sandy wrote these lazy swaps to try and fix `runFixpoint` diverging](https://github.com/polysemy-research/polysemy/pull/194#discussion_r308022711) before I wrote #187. The specialized swaps have since been removed, but this usage remained, because I was unsure if it mattered. I now know it doesn't: `Lazy.WriterT` `Monad` instance always constructs the resulting tuple lazily, so there's no scenario where `swap <$> Lazy.runWriterT (go sem)` would lead to divergence but `(\ ~(a,s) -> (s, a))<$> Lazy.runWriterT (go sem)` wouldn't (the latter being equivalent to the old definition).